### ISSUE TITLE
`used-before-assignment` false positive with walrus

### ISSUE THREAD

**Issue by nickdrozd** [pylint-dev/pylint#9984]
### Bug description

```python
from collections import defaultdict

KEYS: dict[int, dict[int, int]] = defaultdict(dict)

def f(x: int, y: int, z: int) -> int:
    try:
        return (keys := KEYS[x])[y]
    except KeyError:
        keys[y] = z  # [used-before-assignment] FALSE POSITIVE
        return z

print(f(1, 2, 3))
print(f(4, 5, 6))
print(f(4, 5, 6))
```

### Command used

```shell
pylint asdf.py
```

### Pylint output

```python
pylint asdf.py 
************* Module asdf
asdf.py:9:8: E0601: Using variable 'keys' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 5.45/10 (previous run: 5.45/10, +0.00)
```

### Expected behavior

`KEYS[x]` succeeds, and so does `keys := KEYS[x]`. `keys[y]` may fail with `KeyError`, but by that time `keys` is defined, so its use in the `except` block works.

It is important that `KEYS` is `defaultdict`. If it were a regular dictionary, `KEYS[x]` might fail, in which case `keys` would indeed be undefined in the `except` block. Is that something inference can determine? Possibly this should be `possibly-used-before-assignment` instead.

### Pylint version

```shell
pylint 4.0.0-dev0
astroid 3.4.0-dev0
Python 3.12.3 (tags/v3.12.3:f6650f9ad73, May 28 2024, 17:31:51) [GCC 9.4.0]
```

**Comment by jacobtylerwalls** [comment]
> keys[y] may fail with KeyError, but by that time keys is defined, so its use in the except block works.
